### PR TITLE
fix: [CI-9262]: Support for log url from delegate

### PR DESCRIPTION
**Pull request by smjt-h**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52758)

### CODE REVIEW

**Comment by smjt-h** [comment]
trigger uts

**Comment by smjt-h** [comment]
Trigger uts

**Comment by smjt-h** [comment]
trigger uts

**Review by harshjain12 — APPROVED** [review]

File: product/ci/common/external/util.go [hunk 1]
```diff
@@ -485,3 +497,25 @@ func GetStepStrategyIterationsFromEnv() (int, error) {
...
 }
+
+func pingLogService(baseurl string) bool {
+	// Define the URL to ping
+	urlToPing := baseurl + LogHealthEndpoint
+
+	// Send an HTTP GET request to the URL
+	response, err := http.Get(urlToPing)
+	if err != nil {
+		fmt.Println("Error sending GET request:", err)
```
Comment (vistaarjuneja) [inline product/ci/common/external/util.go:508]: use log for errors instead of fmt everywhere

File: product/ci/common/external/util.go [hunk 2]
```diff
@@ -485,3 +497,25 @@ func GetStepStrategyIterationsFromEnv() (int, error) {
...
+		return false
+	}
+	defer response.Body.Close()
+
+	// Check the response status code
+	if response.StatusCode == http.StatusOK {
```
Comment (vistaarjuneja) [inline product/ci/common/external/util.go:514]: this logic should be part of golang client

File: product/ci/common/external/util.go [hunk 3]
```diff
@@ -485,3 +497,25 @@ func GetStepStrategyIterationsFromEnv() (int, error) {
 	}
 	return total, nil
 }
+
+func pingLogService(baseurl string) bool {
+	// Define the URL to ping
+	urlToPing := baseurl + LogHealthEndpoint
```
[the remark this answers is not in the document]
Comment (rutvijmehta-harness) [reply product/ci/common/external/util.go:503]: +1 on this. If custom certs are installed, the Get call will fail.

File: product/ci/common/external/util.go [hunk 4]
```diff
@@ -169,10 +172,19 @@ func GetHTTPRemoteLogger(key string) (*logs.RemoteLogger, error) {
 
 // GetRemoteHTTPClient returns a new HTTP client to talk to log service using information available in env.
 func GetRemoteHTTPClient() (client.Client, error) {
-	l, ok := os.LookupEnv(logSvcEp)
-	if !ok {
-		return nil, fmt.Errorf("log service endpoint variable not set %s", logSvcEp)
+
+	var url string
+	url, ok := os.LookupEnv(delegateLogURLEnabledEnv)
+	if ok && pingLogService(url) {
```
Comment (rutvijmehta-harness) [inline product/ci/common/external/util.go:178]: Optimization: Only ping if `delegateLogURLEnabledEnv != logSvcEp`

We're adding one additional API call everytime the client is created.

**Pull request merged**